modules/calib3d/src/calibinit.cpp

```diff
@@ -191,38 +195,204 @@ static void icvRemoveQuadFromGroup(CvCBQuad **quads, int count, CvCBQuad *q0);
-                CV_MAT_ELEM( *xy, float, i*2+1, 0 ) = CV_MAT_ELEM( *xy, float, i*2+1, 1 ) = CV_MAT_ELEM( *xy, float, i*2+1, 4 ) = 0;
-            CV_MAT_ELEM( *xy, float, i*2, 4 ) = CV_MAT_ELEM( *xy, float, i*2+1, 5 ) = 1;
-            CV_MAT_ELEM( *uv, float, i*2, 0 ) = pts1[j].x;
-            CV_MAT_ELEM( *uv, float, i*2+1, 0 ) = pts1[j].y;
-            i++;
-        }
+      iIdx = i+ii;
+      if (iIdx > 0 && iIdx < 256)
+      {
+        iSmooth += piHist[iIdx];
+      }
+    }
+    piHistSmooth[i] = iSmooth/(2*iWidth+1);
+  }
+  return 0;
+}
+/***************************************************************************************************/
+int icvGradientOfHistogram( const std::vector<int>& piHist, std::vector<int>& piHistGrad )
+{
+  piHistGrad[0] = 0;
+  for ( int i=1; i<255; i++)
+  {
+    piHistGrad[i] = piHist[i-1] - piHist[i+1];
+    if ( abs(piHistGrad[i]) < 100 )
+    {
+      if ( piHistGrad[i-1] == 0)
+        piHistGrad[i] = -100;
+      else
+        piHistGrad[i] = piHistGrad[i-1];
+    }
+  }
+  return 0;
+}
+/***************************************************************************************************/
+bool icvBinarizationHistogramBased( unsigned char* pucImg, int iCols, int iRows )
+{
+  int iMaxPix = iCols*iRows;
+  int iMaxPix1 = iMaxPix/100;
+  const int iNumBins = 256;
+  std::vector<int> piHistIntensity(iNumBins, 0);
+  std::vector<int> piHistSmooth(iNumBins, 0);
+  std::vector<int> piHistGrad(iNumBins, 0);
+  std::vector<int> piAccumSum(iNumBins, 0);
+  std::vector<int> piMaxPos(20, 0);
+  int iThresh = 0;
+  int iIdx;
+  int iWidth = 1;
+
+  icvGetIntensityHistogram( pucImg, iCols, iRows, piHistIntensity );
+
+  // get accumulated sum starting from bright
+  piAccumSum[iNumBins-1] = piHistIntensity[iNumBins-1];
+  for ( int i=iNumBins-2; i>=0; i-- )
+  {
+    piAccumSum[i] = piHistIntensity[i] + piAccumSum[i+1];
+  }
+
+  // first smooth the distribution
+  icvSmoothHistogram( piHistIntensity, piHistSmooth, iWidth );
+
+  // compute gradient
+  icvGradientOfHistogram( piHistSmooth, piHistGrad );
+
+  // check for zeros
+  int iCntMaxima = 0;
+  for ( int i=iNumBins-2; (i>2) && (iCntMaxima<20); i--)
+  {
+    if ( (piHistGrad[i-1] < 0) && (piHistGrad[i] > 0) )
+    {
+      piMaxPos[iCntMaxima] = i;
+      iCntMaxima++;
+    }
+  }
+
+  iIdx = 0;
+  int iSumAroundMax = 0;
+  for ( int i=0; i<iCntMaxima; i++ )
+  {
+    iIdx = piMaxPos[i];
+    iSumAroundMax = piHistSmooth[iIdx-1] + piHistSmooth[iIdx] + piHistSmooth[iIdx+1];
+    if ( iSumAroundMax < iMaxPix1 && iIdx < 64 )
+    {
+      for ( int j=i; j<iCntMaxima-1; j++ )
+      {
+        piMaxPos[j] = piMaxPos[j+1];
+      }
+      iCntMaxima--;
+      i--;
+    }
+  }
+  if ( iCntMaxima == 1)
+  {
+    iThresh = piMaxPos[0]/2;
+  }
+  else if ( iCntMaxima == 2)
+  {
+    iThresh = (piMaxPos[0] + piMaxPos[1])/2;
+  }
+  else // iCntMaxima >= 3
```

Comment:
Case `iCntMaxima == 0` is lost and incorrectly handled by "iCntMaxima >= 3" code.

#11590